### PR TITLE
detect when ties in ranks exist and appropriately warn users and fall back to approx

### LINKED ISSUE #13689: Wilcoxon does not appropriately detect ties when mode=exact.
My issue is about `stats.wilcoxon()`. `stats.wilcoxon` detects when d== 0, and appropriately warns that 'exact' will not work under such a circumstance. However, it reports that this is because of 'ties', and not because of the presence of 0s. 

Exact calculations will also not be accurate when there are ties in rank, but the function does not appropriately detect such scenarios.

R produces appropriate warnings in both circumstances

```
wilcox.test(0:4)

	Wilcoxon signed rank test with continuity correction

data:  0:4
V = 10, p-value = 0.1003
alternative hypothesis: true location is not equal to 0

Warning message:
In wilcox.test.default(0:4) : cannot compute exact p-value with zeroes
> wilcox.test(c(1,1:4))

	Wilcoxon signed rank test with continuity correction

data:  c(1, 1:4)
V = 15, p-value = 0.05791
alternative hypothesis: true location is not equal to 0

Warning message:
In wilcox.test.default(c(1, 1:4)) : cannot compute exact p-value with ties
```

#### Reproducing code example:

```
>>> from scipy import stats as stats
>>> x = [1, 1, 2, 3, 4]
>>> stats.wilcoxon(x, correction=True, mode='exact') ## should produce a warning similar to R and should fall back to approx
WilcoxonResult(statistic=0.0, pvalue=0.0625)
>>> stats.wilcoxon(x, correction=True, mode='approx')
/home/jrr/miniconda3/lib/python3.8/site-packages/scipy/stats/morestats.py:2981: UserWarning: Sample size too small for normal approximation.
  warnings.warn("Sample size too small for normal approximation.")
WilcoxonResult(statistic=0.0, pvalue=0.05790726541729728)
>>> x = [0, 1, 2, 3, 4]
>>> stats.wilcoxon(x, correction=True) ## warns, but calls ties.
/home/jrr/miniconda3/lib/python3.8/site-packages/scipy/stats/morestats.py:2967: UserWarning: Exact p-value calculation does not work if there are ties. Switching to normal approximation.
  warnings.warn("Exact p-value calculation does not work if there are "
WilcoxonResult(statistic=0.0, pvalue=0.10034824646229075)
```

#### Scipy/Numpy/Python version information:
```
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)

1.6.1 1.19.2 sys.version_info(major=3, minor=8, micro=8, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by jreynolds01**
#### Reference issue

Closes #13689 

#### What does this implement/fix?

Fixes appropriate detection of ties and fall back to approx mode.

### CODE REVIEW

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -2966,6 +2966,11 @@ def wilcoxon(x, y=None, zero_method="wilcox", correction=False,
 
     n_zero = np.sum(d == 0)
     if n_zero > 0 and mode == "exact":
+        mode = "approx"
+        warnings.warn("Exact p-value calculation does not work if there are "
+                      "zeros. Switching to normal approximation.")
+
+    if np.unique(abs(d)).size < d.size and mode == "exact":
```
Comment (tylerjereddy) [inline scipy/stats/morestats.py:2973]: Maybe obvious, but just a quick note that these two conditions are never true together in our test suite at the moment (see below from `codecov`). That doesn't mean changes are needed, but maybe useful to know for a `stats` reviewer when they take a look.

image

**Comment by mdhaber** [comment]
This is something about our API that needs to be discussed more generally. @jreynolds01, are you on the [SciPy-dev mailing list](https://mail.python.org/mailman/listinfo/scipy-dev)? A [message](https://mail.python.org/pipermail/scipy-dev/2021-March/024612.html) recently went out about gh-13650. It would be great if you would express your opinion on the mailing list. 

I'm conflicted here. The default method is `'auto'`, and it switches to the asymptotic method when there are ties. That behavior is appropriate. If the user chooses to ignore the documentation (which we _should_ make more explicit) that the exact method does not adjust for ties and override the default by choosing `'exact'`, there is something to be said for allowing that - mostly because in some functions this will eliminate the need to check for ties when the exact method is selected.

**Review by mdhaber — COMMENTED** [review]
Thanks for the contribution @jreynolds01. I took a closer look, though, and I'm afraid there is either a misunderstanding about the test or the variable `d`. 

`wilcoxon` is used to perform a paired-sample test. When only `x` is provided, `x` is the difference between a pair of samples, and `d = x`. When two samples `x` and `y` are provided (where `x[i]` is paired with `y[i]`), `d = x - y`. In either case, `d==0` means that a pair of observations is tied. 

So it is always accurate to detect when `d==0` and give the error message "Exact p-value calculation does not work if there are ties", as in master. This PR would change that message to "Exact p-value calculation does not work if there are zeros". This would be _true_ - although perhaps a little less clear about what the problem really is - when only `x` is provided. But it would be confusing when `x` and `y` are provided, because it is not about either `x` or `y` having a zero element. 

~~I don't think the new check `if np.unique(abs(d)).size < d.size` is appropriate, because the test has no problem when the _difference_ `d` between a pair of samples is not unique.~~ _It does change the null distribution of the test statistic._

Also, there is no guarantee that switching to the normal approximation in case of ties will be any more accurate when the sample size is small. For this test (and others) [Hollander and Wolfe "Nonparametric Statistical Methods"](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=nonparametric+statistical+methods&btnG=) recommends "If there are ties among the _N_ observations, use average ranks to compute _W_, and carry on as in [the case without ties]" (unless the sample is large, in which case it suggests the asymptotic approximation with tie correction). So for small samples there is precedent to use the same null distribution (computed without considering ties) even in the presence of ties. I believe this acceptable because ties tend to reduce the variance of the null distribution, so the returned p-values are conservative (more chance of type-II error, less chance of type I error).

@jreynolds01 perhaps the best thing to do here would be to actually consider ties in the calculation of the null distribution. Would you be interested in extending the method I implemented [here](https://github.com/scipy/scipy/pull/13438#pullrequestreview-577009696) to handle the case of ties?

**Comment by mdhaber** [comment]
This has merge conflicts since the move from `morestats.py` to `_morestats.py`. To avoid additional conflicts with gh-13438, let's close this and consider this conversation over there.
Note that `scipy.stats.permutation_test` can calculate the exact distribution of the test statistic even when there are ties, but if someone is interested in implementing a more efficient algorithm for `wilcoxon` in the presence of ties, that would be the best solution. (E.g., see references linked from [here](https://github.com/scipy/scipy/pull/13438#pullrequestreview-577009696)).
Thanks @jreynolds01!

**Pull request closed without merging**